heron/scheduler-core/src/java/com/twitter/heron/scheduler/UpdateTopologyManager.java

```diff
@@ -0,0 +1,249 @@
+    TopologyAPI.Topology updatedTopology = stateManager.getTopology(topologyName);
+    Map<String, Integer> proposedComponentCounts = proposedPackingPlan.getComponentCounts();
+    return mergeTopology(updatedTopology, proposedComponentCounts);
+  }
+
+  @VisibleForTesting
+  void validateCurrentPackingPlan(PackingPlans.PackingPlan existingProtoPackingPlan,
+                                  String topologyName,
+                                  SchedulerStateManagerAdaptor stateManager) {
+    PackingPlans.PackingPlan foundPackingPlan = stateManager.getPackingPlan(topologyName);
+    assertTrue(foundPackingPlan.equals(existingProtoPackingPlan),
+        "Existing packing plan received does not equal the packing plan found in the state "
+            + "manager. Not updating updatedTopology. Received: %s, Found: %s",
+        existingProtoPackingPlan, foundPackingPlan);
+  }
+
+  @VisibleForTesting
+  static class ContainerDelta {
+    private final Set<PackingPlan.ContainerPlan> containersToAdd;
+    private final Set<PackingPlan.ContainerPlan> containersToRemove;
+
+    @VisibleForTesting
+    ContainerDelta(Set<PackingPlan.ContainerPlan> currentContainers,
+                   Set<PackingPlan.ContainerPlan> proposedContainers) {
+      Set<PackingPlan.ContainerPlan> toAdd = new HashSet<>();
+      for (PackingPlan.ContainerPlan proposedContainerPlan : proposedContainers) {
+        if (!currentContainers.contains(proposedContainerPlan)) {
+          toAdd.add(proposedContainerPlan);
+        }
+      }
+      this.containersToAdd = Collections.unmodifiableSet(toAdd);
+
+      Set<PackingPlan.ContainerPlan> toRemove = new HashSet<>();
+      for (PackingPlan.ContainerPlan curentContainerPlan : currentContainers) {
+        if (!proposedContainers.contains(curentContainerPlan)) {
+          toRemove.add(curentContainerPlan);
+        }
+      }
+      this.containersToRemove = Collections.unmodifiableSet(toRemove);
+    }
+
+    @VisibleForTesting
+    Set<PackingPlan.ContainerPlan> getContainersToRemove() {
+      return containersToRemove;
+    }
+
+    @VisibleForTesting
+    Set<PackingPlan.ContainerPlan> getContainersToAdd() {
+      return containersToAdd;
+    }
+  }
+
+  /**
+   * For each of the components with changed parallelism we need to update the Topology configs to
+   * represent the change.
+   */
+  @VisibleForTesting
+  static TopologyAPI.Topology mergeTopology(TopologyAPI.Topology topology,
+                                            Map<String, Integer> proposedComponentCounts) {
+    TopologyAPI.Topology.Builder builder = TopologyAPI.Topology.newBuilder().mergeFrom(topology);
+    for (String componentName : proposedComponentCounts.keySet()) {
+      Integer parallelism = proposedComponentCounts.get(componentName);
+
+      boolean updated = false;
+      for (TopologyAPI.Bolt.Builder boltBuilder : builder.getBoltsBuilderList()) {
+        if (updateComponent(boltBuilder.getCompBuilder(), componentName, parallelism)) {
```

Comment:
updateComponent will return false under normal circumstances (see the updateComponent javadocs). Because in the protobuf world Bolt and Spout are both Components, but without inheritance, we have to try to update each to see if we have a match. We use true/false so if we get a match on bolt we don't try again on spout.